<h1 style="text-align:center"><b>Data Analysis Project</b></h1>

<center>

### Members of Group 3

Kristian Brilyawan

Farhan Trinanda Putra

Muhammad Tofik Hidayat

Ridho Aryawan

Adelia Januarto

</center>

# **1 Instructions**
___
## About the Company
In 2016, Cyclistic launched a successful bike-share offering. Since then, the program has grown to a fleet of 5,824 bicycles that are geo tracked and locked into a network of 692 stations across Chicago. The bikes can be unlocked from one station and returned to any other station in the system anytime.

>💡 **Cyclistic:** A bike-share program that features more than 5,800 bicycles and 600 docking stations. Cyclistic sets itself apart by also offering reclining bikes, hand tricycles, and cargo bikes, making bike-share more inclusive to people with disabilities and riders who can’t use a standard two-wheeled bike.

## Business Context

Until now, Cyclistic’s marketing strategy relied on building general awareness and appealing to broad consumer segments. One approach that helped make these things possible was the flexibility of its pricing plans: single-ride passes, full-day passes, and annual memberships. Customers who purchase single-ride or full-day passes are referred to as casual riders. Customers who purchase annual memberships are Cyclistic members.

> **Cyclistic’s finance analysts have concluded that annual members are much more profitable than casual riders**. Although the pricing flexibility helps Cyclistic attract more customers, Moreno, the director of marketing, believes that maximizing the number of annual members will be key to future growth. Rather than creating a marketing campaign that targets all-new customers, Moreno believes there is a very good chance to convert casual riders into members.

## Business Task

Moreno and her team are interested in analyzing the Cyclistic historical bike trip data to identify trends. Moreno has set a clear goal: _**Design marketing strategies aimed at converting casual riders into annual members**_. In order to do that, however, the marketing analyst team needs to better understand:

1. How annual members and casual riders differ?
2. Why casual riders would buy a membership?
3. How digital media could affect their marketing tactics?

## Scenario

You are a junior data analyst working in the marketing analyst team at Cyclistic, a bike-share company in Chicago. Therefore, your team wants to understand how casual riders and annual members use Cyclistic bikes differently. From these insights, your team will design a new marketing strategy to convert casual riders into annual members.

The key stakeholders in this project are:

- **Lily Moreno:** The director of marketing and your manager. Moreno is responsible for the development of campaigns and initiatives to promote the bike-share program. These may include email, social media, and other channels.
- **Cyclistic marketing analytics team:** A team of data analysts who are responsible for collecting, analyzing, and reporting data that helps guide Cyclistic marketing strategy.
- **Cyclistic executive team:** The notoriously detail-oriented executive team will decide whether to approve the recommended marketing program.

## Available Data

You will use Cyclistic’s historical trip data to analyze and identify trends. This is public data that you can use to explore how different customer types are using Cyclistic bikes. The data has been made available by Motivate International Inc. under this **[license](https://ride.divvybikes.com/data-license-agreement),** you can download the data [**here**](https://divvy-tripdata.s3.amazonaws.com/index.html).

<center>

<table>
  <tr>
    <td><b>Column</b></td>
    <td><b>Description</b></td>
  </tr>
  <tr>
    <td>ride_id</td>
    <td>The identification number for ride from start station to end station in given time.</td>
  </tr>
  <tr>
    <td>rideable_type</td>
    <td>Cyclistic company bike type, there are 3 bike type classic bike, docked bike, and electric bike.</td>
  </tr>
  <tr>
    <td>started_at</td>
    <td>Date and time when the ride begin.</td>
  </tr>
  <tr>
    <td>ended_at</td>
    <td>Date and time when the ride end.</td>
  </tr>
  <tr>
    <td>day</td>
    <td>Day name when the ride begin.</td>
  </tr>
  <tr>
    <td>start_station_name</td>
    <td>The name of the station where the ride begin.</td>
  </tr>
  <tr>
    <td>start_station_id</td>
    <td>The id of start station.</td>
  </tr>
  <tr>
    <td>end_station_name</td>
    <td>The name of the station where the ride end.</td>
  </tr>
  <tr>
    <td>end_station_id</td>
    <td>The id of end station.</td>
  </tr>
  <tr>
    <td>member_casual</td>
    <td>The riders type in Cyclistic company.</td>
  </tr>
  <tr>
    <td>start_lat</td>
    <td>Latitude of the start station.</td>
  </tr>
  <tr>
    <td>start_lng</td>
    <td>Longitude of the start station.</td>
  </tr>
  <tr>
    <td>end_lat</td>
    <td>Latitude of the end station.</td>
  </tr>
  <tr>
    <td>end_lng</td>
    <td>Longitude of the end station.</td>
  </tr>
</table>


</center>

# **2 Data Cleaning**
___

<b>Procedures</b>
1. Import and concat dataframes
2. Perform basic sanity checking
    - Check duplicated entry
    - Check outliers
    - Check inconsistent formats
    - Check missing values

In [1]:
# import libraries
import pandas as pd                             # for data wrangling
import numpy as np                              # for numeric operations
import altair as alt                            # for data visualization
alt.data_transformers.disable_max_rows()        # for disabling max row num in Altair

DataTransformerRegistry.enable('default')

## Data Imports and Concatenation

In [53]:
# import data: Jul 2023
df1 = pd.read_csv('../data/202307-divvy-tripdata.csv', parse_dates=True)

# import data: Aug 2023
df2 = pd.read_csv('../data/202308-divvy-tripdata.csv', parse_dates=True)

# import data: Sep 2023
df3 = pd.read_csv('../data/202309-divvy-tripdata.csv', parse_dates=True)

In [54]:
# append three DataFrames
df = pd.concat([df1, df2, df3], ignore_index=True)

# check top-5 rows
df.head()

ride_id  rideable_type           started_at             ended_at  \
0  9340B064F0AEE130  electric_bike  2023-07-23 20:06:14  2023-07-23 20:22:44   
1  D1460EE3CE0D8AF8   classic_bike  2023-07-23 17:05:07  2023-07-23 17:18:37   
2  DF41BE31B895A25E   classic_bike  2023-07-23 10:14:53  2023-07-23 10:24:29   
3  9624A293749EF703  electric_bike  2023-07-21 08:27:44  2023-07-21 08:32:40   
4  2F68A6A4CDB4C99A   classic_bike  2023-07-08 15:46:42  2023-07-08 15:58:08   

         start_station_name start_station_id  \
0     Kedzie Ave & 110th St            20204   
1   Western Ave & Walton St     KA1504000103   
2   Western Ave & Walton St     KA1504000103   
3  Racine Ave & Randolph St            13155   
4     Clark St & Leland Ave     TA1309000014   

                      end_station_name end_station_id  start_lat  start_lng  \
0  Public Rack - Racine Ave & 109th Pl            877  41.692406 -87.700905   
1            Milwaukee Ave & Grand Ave          13033  41.898418 -87.686596   
2               Damen Ave & Pierce Ave   TA1305000041  41.898418 -87.686596   
3              Clinton St & Madison St   TA1305000032  41.884112 -87.656943   
4                      Montrose Harbor   TA1308000012  41.967088 -87.667291   

     end_lat    end_lng member_casual  
0  41.694835 -87.653041        member  
1  41.891578 -87.648384        member  
2  41.909396 -87.677692        member  
3  41.882752 -87.641190        member  
4  41.963982 -87.638181        member

In [55]:
# get the df dimensions
df1_shape = df1.shape
df2_shape = df2.shape
df3_shape = df3.shape
df_shape = df.shape
print(f'''
- Shape of df1: {df1_shape}
- Shape of df2: {df2_shape}
- Shape of df3: {df3_shape}
- Shape of concatenated DataFrames: {df_shape}''')


- Shape of df1: (767650, 13)
- Shape of df2: (771693, 13)
- Shape of df3: (666371, 13)
- Shape of concatenated DataFrames: (2205714, 13)


<span style='color:red'><b>Comments</b></span>: After the concatenation, the rows of the merged dataframe are 2,205,714

In [56]:
# get info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2205714 entries, 0 to 2205713
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 218.8+ MB


<span style='color:red'><b>Comments</b></span>: The started_at and ended_at are supposed to be datetime, not objects so they need to be converted into datetime.

In [57]:
# dtype conversion
df['started_at'] = pd.to_datetime(df['started_at'])
df['ended_at'] = pd.to_datetime(df['ended_at'])

## Data Sanity Check
- Duplicated entry check
- Inconsistent format check
- Missing data check

### Check Duplicates

In [58]:
# check duplicates
df.duplicated().sum()

0

<span style='color:red'><b>Comments</b></span>: No duplicates found. No treatment for duplicated values will be performed.

### Check Inconsistent Format

In [59]:
# check length of end_station_id > 8
end_station_id_cnt = df['end_station_id'].value_counts().reset_index(name='count').rename(columns={'index':'end_station_id'})
(end_station_id_cnt[end_station_id_cnt['end_station_id'].str.len() > 8]).nlargest(columns='count', n=10)

end_station_id  count
4    TA1308000001  15463
6    TA1307000039  13574
7    TA1308000050  13082
9    KA1503000043  11759
11   KA1504000135  11380
12   TA1308000012  10899
13   TA1307000134  10848
15   TA1309000039  10646
18   TA1307000126  10148
19   TA1309000049   9869

In [60]:
# sort end_station_id in desc order
end_station_id_cnt.sort_values(by='end_station_id', key=lambda x: x.str.len(), ascending=False).head(10)

end_station_id  count
821  Hubbard Bike-checking (LBS-WH-TEST)     28
269                        chargingstx07   2423
255                        chargingstx06   2592
606                         KA1503000021    104
210                         TA1305000035   3185
215                         TA1305000005   3125
214                         TA1309000043   3134
213                         TA1307000061   3136
212                         TA1309000059   3140
449                         TA1309000067    592

<span style='color:red'><b>Comments</b></span>: The formats for `end_station_id` are inconsistent. Many station ID names consist of combination between two letters followed by 10 digits of numbers but some station ID names only consist of numbers only.

### Check Missing Values

In [61]:
# check missing values
missingValues = (df.isna().mean() * 100).to_frame().rename(columns={'index':'columns', 0:'percent'})
missingValues['count'] = df.isna().sum()
missingValues['percent'] = round(missingValues['percent'], 2).astype('str') + '%'
missingValues = missingValues.reset_index(names='column')
missingValues

column percent   count
0              ride_id    0.0%       0
1        rideable_type    0.0%       0
2           started_at    0.0%       0
3             ended_at    0.0%       0
4   start_station_name  15.56%  343174
5     start_station_id  15.56%  343174
6     end_station_name  16.46%  363163
7       end_station_id  16.46%  363163
8            start_lat    0.0%       0
9            start_lng    0.0%       0
10             end_lat   0.15%    3349
11             end_lng   0.15%    3349
12       member_casual    0.0%       0

In [62]:
na_df = (df.isna().mean()*100).reset_index(name='count')
na_df['count'] = round(na_df['count'], 2).astype('str') + '%'
na_df

index   count
0              ride_id    0.0%
1        rideable_type    0.0%
2           started_at    0.0%
3             ended_at    0.0%
4   start_station_name  15.56%
5     start_station_id  15.56%
6     end_station_name  16.46%
7       end_station_id  16.46%
8            start_lat    0.0%
9            start_lng    0.0%
10             end_lat   0.15%
11             end_lng   0.15%
12       member_casual    0.0%

In [63]:
na_df = (df.isna().mean()*100).reset_index(name='count')
na_df['count'] = round(na_df['count'], 2)

bar = alt.Chart(na_df).mark_bar().encode(
    alt.X('count', title='Percentage'),
    alt.Y('index', title=None, sort='-x')
)
text = bar.mark_text(
    align='center',
    baseline='middle',
    dx= 15
).encode(
    alt.Text('count', format='.2f') 
)
(bar + text).properties(
    title=alt.Title('Station names, ids, and coordinates contain missing values',
                    anchor='start',
                    fontSize=20,
                    offset=20),
                    width=600, height=400
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.LayerChart(...)

<span style='color:red'><b>Comments</b></span>: 
- Six columns contain missing values, and four of them have a huge number of missing values (far beyond 2%). Therefore, removing observations with missing values is not recommended. Some possible solutions are (1) <span style='color:orange'>data imputation</span>, e.g., using central tendency measures, or (2) <span style='color:orange'>flagging</span> (replacing missing values with an arbitrary number or categor, e.g., 'Unknown'). 
- Missing value removal in columns `start_station_name`, `start_station_id`, `end_station_name`, and `end_station_id` is not possible to perform since the number of missing values is significant. 
- Flagging can be done as data imputationt technique for `start_station_name` and `end_station_name` 
- Missing values in `end_lat` and `end_lng`, however, can be removed as the number is less than 2%.

In [64]:
# check profiles of cols with missing values
df[df.start_station_name.isna()].sample(5)

ride_id  rideable_type          started_at  \
1984713  7F818F8FC18F8821  electric_bike 2023-09-17 06:17:50   
2075097  EF8961B6B4C82235  electric_bike 2023-09-07 15:53:35   
754486   A9506D21F3FA0082  electric_bike 2023-07-27 19:54:41   
738539   21BE9766E38BF3A1  electric_bike 2023-07-09 10:08:01   
1442672  7E62B3306AC2564D  electric_bike 2023-08-08 11:19:02   

                   ended_at start_station_name start_station_id  \
1984713 2023-09-17 06:29:31                NaN              NaN   
2075097 2023-09-07 16:04:56                NaN              NaN   
754486  2023-07-27 20:13:19                NaN              NaN   
738539  2023-07-09 10:28:33                NaN              NaN   
1442672 2023-08-08 11:24:22                NaN              NaN   

                end_station_name end_station_id  start_lat  start_lng  \
1984713                      NaN            NaN      41.94     -87.64   
2075097  Clark St & Armitage Ave          13146      41.91     -87.67   
754486                       NaN            NaN      41.88     -87.65   
738539                       NaN            NaN      41.94     -87.64   
1442672                      NaN            NaN      41.85     -87.61   

           end_lat    end_lng member_casual  
1984713  41.960000 -87.650000        member  
2075097  41.918306 -87.636282        member  
754486   41.890000 -87.620000        member  
738539   41.920000 -87.630000        casual  
1442672  41.840000 -87.600000        casual

<span style='color:red'><b>Comments</b></span>: One possible solution for resolving the missing values in the `start_station_name` is by using latitude and longitude information. The latitude and longitude can pinpoint a certain location. For example, using `start_lat` and `start_lng` to get the `start_station_name`.

In [65]:
df[(df['start_station_name'].isna()) & df['start_lat'].notna()].loc[:, ['start_station_name', 'start_lat', 'start_lng']].sample(20, random_state=123)

start_station_name  start_lat  start_lng
2185762                NaN      41.95     -87.67
2043824                NaN      41.87     -87.62
2180299                NaN      41.89     -87.63
759125                 NaN      41.89     -87.63
721684                 NaN      41.92     -87.67
223727                 NaN      41.80     -87.60
1290680                NaN      41.91     -87.68
1519815                NaN      41.89     -87.63
363097                 NaN      41.89     -87.63
245282                 NaN      41.87     -87.68
900493                 NaN      41.90     -87.64
1489216                NaN      41.92     -87.72
1437508                NaN      41.92     -87.70
1987090                NaN      41.91     -87.68
1144323                NaN      41.83     -87.63
289341                 NaN      41.93     -87.72
1290578                NaN      41.94     -87.67
312777                 NaN      41.91     -87.63
1525536                NaN      41.88     -87.65
1376808                NaN      41.96     -87.63

# **3 Data Preprocessing**
___

In [18]:
# replace 'member' with 'subscriber' in member_casual column
df['member_casual'].replace({'member': 'subscriber'}, inplace=True)

In [19]:
# add new columns
df['start_hour'] = df['started_at'].dt.hour
df['start_day'] = df['started_at'].dt.day_name()
df['day_of_week'] = (df['started_at'].dt.dayofweek + 1)
df['day_type'] = df['day_of_week'].map({1: 'weekday', 2: 'weekday', 3: 'weekday', 
                                        4: 'weekday', 5: 'weekday', 6: 'weekend', 7:'weekend'})
df['month'] = df['started_at'].dt.month
df['year'] = df['started_at'].dt.year
df['trip_duration'] = ((df['ended_at'] - df['started_at']).dt.total_seconds()/60).round(2)
df['duration(h)'] = (df['ended_at'] - df['started_at']).dt.total_seconds() / 3600

tarif = {
    ('classic_bike', 'subscriber'): 0.18,
    ('classic_bike', 'casual'): 0.18,
    ('docked_bike', 'subscriber'): 0.29,
    ('docked_bike', 'casual'): 0.44,
    ('electric_bike', 'subscriber'): 0.18,
    ('electric_bike', 'casual'): 0.44,
}

df['charge($)'] = df.apply(lambda df: tarif[(df['rideable_type'], df['member_casual'])] * (df['trip_duration']), axis=1)

df['customer_stage'] = np.where((df['trip_duration'] <= df['trip_duration'].median()) & 
                                (df['member_casual'] == 'casual'), 'aware',
                                  np.where((df['trip_duration'] >= df['trip_duration'].median()) & 
                                           (df['member_casual'] == 'casual'), 'consider', 'convert'))
df.head()

ride_id  rideable_type          started_at            ended_at  \
0  9340B064F0AEE130  electric_bike 2023-07-23 20:06:14 2023-07-23 20:22:44   
1  D1460EE3CE0D8AF8   classic_bike 2023-07-23 17:05:07 2023-07-23 17:18:37   
2  DF41BE31B895A25E   classic_bike 2023-07-23 10:14:53 2023-07-23 10:24:29   
3  9624A293749EF703  electric_bike 2023-07-21 08:27:44 2023-07-21 08:32:40   
4  2F68A6A4CDB4C99A   classic_bike 2023-07-08 15:46:42 2023-07-08 15:58:08   

         start_station_name start_station_id  \
0     Kedzie Ave & 110th St            20204   
1   Western Ave & Walton St     KA1504000103   
2   Western Ave & Walton St     KA1504000103   
3  Racine Ave & Randolph St            13155   
4     Clark St & Leland Ave     TA1309000014   

                      end_station_name end_station_id  start_lat  start_lng  \
0  Public Rack - Racine Ave & 109th Pl            877  41.692406 -87.700905   
1            Milwaukee Ave & Grand Ave          13033  41.898418 -87.686596   
2               Damen Ave & Pierce Ave   TA1305000041  41.898418 -87.686596   
3              Clinton St & Madison St   TA1305000032  41.884112 -87.656943   
4                      Montrose Harbor   TA1308000012  41.967088 -87.667291   

   ...  start_hour  start_day day_of_week  day_type month  year trip_duration  \
0  ...          20     Sunday           7   weekend     7  2023         16.50   
1  ...          17     Sunday           7   weekend     7  2023         13.50   
2  ...          10     Sunday           7   weekend     7  2023          9.60   
3  ...           8     Friday           5   weekday     7  2023          4.93   
4  ...          15   Saturday           6   weekend     7  2023         11.43   

   duration(h)  charge($)  customer_stage  
0     0.275000     2.9700         convert  
1     0.225000     2.4300         convert  
2     0.160000     1.7280         convert  
3     0.082222     0.8874         convert  
4     0.190556     2.0574         convert  

[5 rows x 23 columns]

In [20]:
# drop missing values in 'end_lat', 'end_lng'
df.dropna(subset=['end_lat', 'end_lng'], inplace=True)

In [21]:
# fill in missing values

# initiate counter to count "unknown"
unknown_counter = 1

# initiate dict to track 'unknown' based on 'start_lat' dan 'start_lng'
unknown_types = {}

# iterate over each row of df
for index, row in df.iterrows():
    start_lat = row['start_lat']
    start_lng = row['start_lng']
    
    # check if 'start_station_name' value is NaN
    if pd.isna(row['start_station_name']):
        # create unique key based on start_lat and start_lng
        key = (start_lat, start_lng)
        
        # check if combination already in dict unknown_types
        if key in unknown_types:
            # use type unknown if already exists 
            unknown_type = unknown_types[key]
        else:
            # use new "unknown" and update dict
            unknown_type = f'unknown_{unknown_counter}'
            unknown_types[key] = unknown_type
            unknown_counter += 1
        
        # fill in missing values using unknown
        df.at[index, 'start_station_name'] = unknown_type

# iteerate over each row
for index, row in df.iterrows():
    end_lat = row['end_lat']
    end_lng = row['end_lng']
    
    # chck if start_station_name is NaN
    if pd.isna(row['end_station_name']):
        # create unique key according to combination start_lat and start_lng
        key = (end_lat, end_lng)
        
        # check if combination already exists in dict unknown_types
        if key in unknown_types:
            # use the existing unknown type
            unknown_type = unknown_types[key]
        else:
            # use new unknown type and update the dict
            unknown_type = f'unknown_{unknown_counter}'
            unknown_types[key] = unknown_type
            unknown_counter += 1
        
        # impute missing values using appropriate unknown
        df.at[index, 'end_station_name'] = unknown_type

In [22]:
# recheck missing values
(df.isna().mean()*100).reset_index(name='percentage')

index  percentage
0              ride_id    0.000000
1        rideable_type    0.000000
2           started_at    0.000000
3             ended_at    0.000000
4   start_station_name    0.000000
5     start_station_id   15.582067
6     end_station_name    0.000000
7       end_station_id   16.342432
8            start_lat    0.000000
9            start_lng    0.000000
10             end_lat    0.000000
11             end_lng    0.000000
12       member_casual    0.000000
13          start_hour    0.000000
14           start_day    0.000000
15         day_of_week    0.000000
16            day_type    0.000000
17               month    0.000000
18                year    0.000000
19       trip_duration    0.000000
20         duration(h)    0.000000
21           charge($)    0.000000
22      customer_stage    0.000000

<span style='color:red'><b>Comments</b></span>: 
- As shown previously, majority of columns with missing values (start_station_name, start_station_id, end_station_name, end_station_id) are objects. And because the percentages of missing values are far larger than the widely accepted cut-off (2%), missing value removal was not performed. Instead, we carried out data imputation using the most frequent values in each row (**modes**).
- Additionally, we dropped missing values end_lat, and end_lng because the percentage of missingness is lower than 2% of the total rows in dataset.

In [23]:
# drop unnecessary columns: ride_id, start_station_id, end_station_id
df.drop(['ride_id', 'start_station_id', 'end_station_id'], inplace=True, axis=1)

# check top-5 rows
df.head()

rideable_type          started_at            ended_at  \
0  electric_bike 2023-07-23 20:06:14 2023-07-23 20:22:44   
1   classic_bike 2023-07-23 17:05:07 2023-07-23 17:18:37   
2   classic_bike 2023-07-23 10:14:53 2023-07-23 10:24:29   
3  electric_bike 2023-07-21 08:27:44 2023-07-21 08:32:40   
4   classic_bike 2023-07-08 15:46:42 2023-07-08 15:58:08   

         start_station_name                     end_station_name  start_lat  \
0     Kedzie Ave & 110th St  Public Rack - Racine Ave & 109th Pl  41.692406   
1   Western Ave & Walton St            Milwaukee Ave & Grand Ave  41.898418   
2   Western Ave & Walton St               Damen Ave & Pierce Ave  41.898418   
3  Racine Ave & Randolph St              Clinton St & Madison St  41.884112   
4     Clark St & Leland Ave                      Montrose Harbor  41.967088   

   start_lng    end_lat    end_lng member_casual  start_hour start_day  \
0 -87.700905  41.694835 -87.653041    subscriber          20    Sunday   
1 -87.686596  41.891578 -87.648384    subscriber          17    Sunday   
2 -87.686596  41.909396 -87.677692    subscriber          10    Sunday   
3 -87.656943  41.882752 -87.641190    subscriber           8    Friday   
4 -87.667291  41.963982 -87.638181    subscriber          15  Saturday   

   day_of_week day_type  month  year  trip_duration  duration(h)  charge($)  \
0            7  weekend      7  2023          16.50     0.275000     2.9700   
1            7  weekend      7  2023          13.50     0.225000     2.4300   
2            7  weekend      7  2023           9.60     0.160000     1.7280   
3            5  weekday      7  2023           4.93     0.082222     0.8874   
4            6  weekend      7  2023          11.43     0.190556     2.0574   

  customer_stage  
0        convert  
1        convert  
2        convert  
3        convert  
4        convert

# **4 Exploratory Data Analysis**
___

## `Question 1`: How annual members and casual riders differ?
- Berapa banyak riders di setiap membership? 

In [24]:
counts = df['member_casual'].value_counts().reset_index()
counts['percent'] = (counts['member_casual']/counts['member_casual'].sum() * 100).round(2)
counts

index  member_casual  percent
0  subscriber        1301125    59.08
1      casual         901240    40.92

In [25]:
base = alt.Chart(counts).mark_bar().encode(
    x=alt.X('index', sort='-y', title='Membership Type', axis=alt.Axis(labelAngle=0)),
    y=alt.Y('member_casual', title='Count').axis(format=".3s"),
    color=alt.Color('index', scale=alt.Scale(range=['orange', 'blue']), legend=None),
    tooltip=['index', 'member_casual']
)

text = base.mark_text(
    align='right',
    baseline='middle',
    fontSize=12
)

(base + text).properties(
    title=alt.Title('Majority of riders are annual members',
                    subtitle='Comparison between membership types',
                    anchor='start',
                    fontSize=18,
                    subtitleFontSize=14,
                    offset=20,),
                        width=600,
                        height=400
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.LayerChart(...)

In [26]:
base = alt.Chart(counts).mark_bar().encode(
    x=alt.X('index', sort='-y', title='Membership Type', 
            axis=alt.Axis(labelAngle=0)),
    y=alt.Y('member_casual', title='Count').axis(format=".3s"),
    color=alt.Color('index', scale=alt.Scale(range=['orange', 'blue']), legend=None),
    tooltip=['index', 'member_casual']
)

text = base.mark_text(
    align='center',
    baseline='middle',
    dy=-10,
    fontSize=12
)

(base + text).properties(
    title=alt.Title('Majority of riders are annual members',
                    subtitle='Comparison between membership types',
                    anchor='start',
                    fontSize=18,
                    subtitleFontSize=14,
                    offset=20,),
                        width=600,
                        height=400
).encode(
    text=alt.Text('member_casual', format=".3s"),
    x=alt.X('index', sort='-y', title='Membership Type', axis=alt.Axis(labelAngle=0, labelFontSize=12)),
    y=alt.Y('member_casual', title='Count').axis(format=".3s")
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.LayerChart(...)

- Jam berapa annual member dan casual riders biasa menggunakan layanan? 

In [27]:
# Filter data for annual member and casual riders
annual_member_trips = df[df['member_casual']  == 'subscriber']
casual_rider_trips = df[df['member_casual'] == 'casual']

# Count the number of trips starting at each hour
annual_member_counts = annual_member_trips['start_hour'].value_counts().sort_index().reset_index()
casual_rider_counts = casual_rider_trips['start_hour'].value_counts().sort_index().reset_index()

# Rename columns
annual_member_counts.columns = ['start_hour', 'count']
casual_rider_counts.columns = ['start_hour', 'count']

# annual member line
annual_member_chart = alt.Chart(annual_member_counts).mark_line(color='blue').encode(
    x=alt.X('start_hour:Q', title='Start Hour', scale=alt.Scale(domain=(0, 24))).axis(format=".3s"),
    y=alt.Y('count:Q', title='Count'),
    tooltip=['start_hour:Q', 'count:Q']
).properties(
    width=600,
    height=400
)

# casual rider line
casual_rider_chart = alt.Chart(casual_rider_counts).mark_line(color='orange').encode(
    x=alt.X('start_hour:Q', title='Start Hour', scale=alt.Scale(domain=(0, 24))).axis(format=".3s"),
    y='count:Q',
    tooltip=['start_hour:Q', 'count:Q']
).properties(
    width=600,
    height=400
)

# annual member text annotation
annual_member_annotation = alt.Chart(annual_member_counts.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=-25,  
    dy=-35,  
    color='blue',  
    fontSize=12
).encode(
    x=alt.X('start_hour:Q', stack='zero'),
    y=alt.Y('count:Q').axis(format=".3s"),
    text=alt.value('Subscribers')
)

# casual member text annotation
casual_rider_annotation = alt.Chart(casual_rider_counts.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=-5, 
    dy=0,  
    color='orange',  
    fontSize=12
).encode(
    x=alt.X('start_hour:Q', stack='zero'),
    y=alt.Y('count:Q').axis(format=".3s"),
    text=alt.value('Casual Rider')
)

# show combined elements
(annual_member_chart + casual_rider_chart + annual_member_annotation + casual_rider_annotation).properties(
    title=alt.Title('Subscribers are more active before and after work time', 
                    subtitle='Comparison of start hours between membership types',
                    anchor='start',
                    fontSize=18,
                    subtitleFontSize=14,
                    offset=20)
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.LayerChart(...)

- Hari apa saja annual member dan casual riders biasa menggunakan layanan? 

In [28]:
day_membership = pd.crosstab(index=df['day_of_week'], columns=df['member_casual']).reset_index(names='day_of_week')
day_membership['day_of_week'] = day_membership['day_of_week'].astype('str')
day_membership

member_casual day_of_week  casual  subscriber
0                       1  101205      168489
1                       2  105920      196252
2                       3   99716      197280
3                       4  111456      204737
4                       5  139011      193947
5                       6  194318      189205
6                       7  149614      151215

In [29]:
casual_line = alt.Chart(day_membership).mark_line(color='orange').encode(
    alt.X('day_of_week', title='Days of Week', axis=alt.Axis(labelAngle=0)),
    alt.Y('casual', title='Count').axis(format='s')
)

member_line = alt.Chart(day_membership).mark_line(color='blue').encode(
    alt.X('day_of_week'),
    alt.Y('subscriber')
)

# annual member text annotation
annual_member_annotation = alt.Chart(day_membership.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=0,  
    dy=-35,  
    color='blue',  
    fontSize=12
).encode(
    x=alt.X('day_of_week:N'),
    y=alt.Y('subscriber:Q'),
    text=alt.value('Member')
)

# casual member text annotation
casual_rider_annotation = alt.Chart(day_membership.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=-5, 
    dy=0,  
    color='orange',  
    fontSize=12
).encode(
    x=alt.X('day_of_week:N'),
    y=alt.Y('casual:Q'),
    text=alt.value('Casual Rider')
)

(casual_line + member_line + annual_member_annotation + casual_rider_annotation).properties(
    title=alt.Title('Both number of members and casual riders drop in weekends',
                    subtitle='Bike usage by day of the week',
                    fontSize=20,
                    subtitleFontSize=14,
                    anchor='start',
                    offset=20),
                    width=600,height=400
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.LayerChart(...)

- Bagaimana trend penggunaan sepeda di setiap membership?

In [30]:
member_mth = pd.crosstab(index=df['month'], columns=df['member_casual']).reset_index()
casual = alt.Chart(member_mth).mark_line(color='orange').encode(
    x=alt.X('month:N', title='Month', axis=alt.Axis(labelAngle=0)),
    y=alt.Y('casual', title='Count').axis(format=".3s")
).properties(
    width=600,
    height=400
)
member = alt.Chart(member_mth).mark_line(color='blue').encode(
    x=alt.X('month:N'),
    y=alt.Y('subscriber:Q')
).properties(
    width=600,
    height=400
)

# annual member text annotation
annual_member_annotation = alt.Chart(member_mth.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=0,  
    dy=-25,  
    color='blue',  
    fontSize=12
).encode(
    x=alt.X('month:N'),
    y=alt.Y('subscriber:Q'),
    text=alt.value('Subscriber')
)

# casual rider text annotation
casual_annotation = alt.Chart(member_mth.tail(1)).mark_text(
    align='right',
    baseline='middle',
    dx=0,  
    dy=-23,  
    color='orange',  
    fontSize=12
).encode(
    x=alt.X('month:N'),
    y=alt.Y('casual:Q'),
    text=alt.value('Casual Rider')
)

(casual + member + annual_member_annotation + casual_annotation).properties(
    title=alt.Title('Annual members ride bikes far more often throughout Q3',
                    subtitle='Comparison between membership types between Jul and Sep',
                    anchor='start',
                    fontSize=18,
                    subtitleFontSize=14,
                    offset=20,),
                        width=600,
                        height=400
).configure_axis(
    grid=True,
    labelFontSize=12
).configure_view(
    stroke="#ddd"
)

alt.LayerChart(...)

- Jenis sepeda apa yang sering digunakan? 

In [31]:
# prepare data frame
member_bike_type = (df.groupby(['rideable_type','member_casual'])
                    .agg(func={'member_casual':'count'})
                        .reset_index(names=['rideable_type', 'membership'])
                            .rename(columns={'member_casual': 'count'}))
# show data frame
member_bike_type

rideable_type  membership   count
0   classic_bike      casual  421907
1   classic_bike  subscriber  682285
2    docked_bike      casual   33303
3  electric_bike      casual  446030
4  electric_bike  subscriber  618840

In [32]:
alt.Chart(member_bike_type).mark_bar().encode(
    x=alt.X('rideable_type', title='Bike type', sort='-y', axis=alt.Axis(labelAngle=0)),
    y=alt.Y('count', title='Count').axis(format='s'),
    tooltip=alt.Tooltip(['rideable_type', 'count']),
    color=alt.Color('membership', 
                    scale=alt.Scale(range=['orange', 'blue']), 
                    legend=alt.Legend(title='Membership Type')),
).properties(
    title=alt.Title(
        "Annual members have no preference on docked bikes",
        subtitle='Bike type counts based on membership',
        anchor='start',
        fontSize=20,
        subtitleFontSize=14,
        offset=20,
    ),
    width=600, 
    height=400,
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
).configure_legend(
    labelFontSize=12
)

alt.Chart(...)

- Berapa lama penggunaan sepeda pada setiap jenis membership?

In [33]:
membership_duration = df.groupby('member_casual').agg(func={'trip_duration':'mean'}).reset_index()
membership_duration

member_casual  trip_duration
0        casual      22.029477
1    subscriber      13.021409

In [37]:
# get statistical summary for trip_duration
df[df['member_casual'] == 'casual'].trip_duration.describe().round(3)

count    901240.000
mean         22.029
std          45.866
min         -49.530
25%           7.250
50%          12.870
75%          23.670
max        6891.220
Name: trip_duration, dtype: float64

<span style='color:red'><b>Comment</b></span>: 
- The data frame `membership_duration` (a subset of the full data frame) contains unexpected values. In reality, it is not possible for trip duration (subctracted start time from the end time of each trip) to have < 0 min. Even when the customers decided not to use the bike, the value should be at least 0.
- As a result, we will remove the records (trip duration < 0 from the dataset)

In [38]:
# remove illogical trip duration
df = df.drop(df[df['trip_duration'] < 0].index)

In [39]:
alt.Chart(membership_duration).mark_bar().encode(
    x=alt.X('member_casual', title='Membership Type', sort='-y', 
            axis=alt.Axis(labelAngle=0, labelFontSize=12)),
    y=alt.Y('trip_duration', title='Trip Duration (in mins)', axis=alt.Axis(labelFontSize=12)),
    tooltip=alt.Tooltip(['member_casual', 'trip_duration']),
    color=alt.Color('member_casual', 
                    scale=alt.Scale(range=['orange', 'blue']), 
                    legend=None),
).properties(
    title=alt.Title(
        "Casual members have longer trip durations",
        subtitle='Trip Duration Based on Membership Types',
        anchor='start',
        fontSize=18,
        subtitleFontSize=14,
        offset=20,
    ),
    width=600, 
    height=400,
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.Chart(...)

- Bagaimana tren lokasi peminjaman sepeda berdasarkan membership? 

In [50]:
top_station_casual_start = df[df['member_casual'] == 'casual']['start_station_name'].value_counts().reset_index(name='count').head(10)
top_station_member_start = df[df['member_casual'] == 'subscriber']['start_station_name'].value_counts().reset_index(name='count').head(10)
top_station_casual_end = df[df['member_casual'] == 'casual']['end_station_name'].value_counts().reset_index(name='count').head(10)
top_station_member_end = df[df['member_casual'] == 'subscriber']['end_station_name'].value_counts().reset_index(name='count').head(10)

# Charts for top start stations
chart_top_start_casual = alt.Chart(top_station_casual_start).mark_bar(color='orange').encode(
    x=alt.X('count', title='Count', scale=alt.Scale(domain=(0, 25000)), axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    y=alt.Y('index', title=None, sort='-x')
).properties(
    title=alt.Title("Casual Members Start Stations",
                    anchor='start'),
    width=400, height=250,
)

chart_top_start_member = alt.Chart(top_station_member_start).mark_bar(color='blue').encode(
    x=alt.X('count', title='Count', scale=alt.Scale(domain=(0, 25000)), axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    y=alt.Y('index', title=None, sort='-x')
).properties(
    title=alt.Title("Annual Members Start Stations",
                    anchor='start'),
    width=400, height=250,
)

# Charts for top end stations
chart_top_end_casual = alt.Chart(top_station_casual_end).mark_bar(color='orange').encode(
    x=alt.X('count', title='Count', scale=alt.Scale(domain=(0, 25000)), axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    y=alt.Y('index', title=None, sort='-x')
).properties(
    title=alt.Title("Casual Members Stop Stations",
                    anchor='start'),
    width=400, height=250,
)

chart_top_end_member = alt.Chart(top_station_member_end).mark_bar(color='blue').encode(
    x=alt.X('count', title='Count', scale=alt.Scale(domain=(0, 25000)), axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    y=alt.Y('index', title=None, sort='-x')
).properties(
    title=alt.Title("Annual Members Stop Stations",
                    anchor='start'),
    width=400, height=250,
)

# Combine charts
combined_chart = (chart_top_start_casual | chart_top_start_member) & (chart_top_end_casual | chart_top_end_member)

combined_chart.properties(
    title=alt.Title('Annual and casual members have different preferences on Start and End stations',
                    subtitle='Top-10 Stations by Membership Type',
                    anchor='start',
                    fontSize=20,
                    subtitleFontSize=14,
                    offset=20)
).configure_axis(
    grid=False,
    labelFontSize=12
).configure_view(
    stroke=None
)

alt.VConcatChart(...)

## `Question 2`: Why casual riders would buy a membership?

- Berapa pricing di setiap membership?

In [41]:
avg_pricing = df.groupby(['member_casual', 'day_of_week'])['charge($)'].mean().reset_index()

In [42]:
alt.Chart(avg_pricing).mark_line().encode(
    alt.X('day_of_week:N', title='Day of Week', axis=alt.Axis(labelAngle=0)),
    alt.Y('charge($)', title='Price (in USD)').axis(format='s'),
    tooltip=['member_casual', 'charge($)'],
    color=alt.Color('member_casual', scale=alt.Scale(range=['orange', 'blue']),
                    legend=alt.Legend(title='Membership Types')),
).properties(
    title=alt.Title(
        "The avg pricing of casual membership is more expensive",
        subtitle='Comparison of pricing between membership types',
        anchor='start',
        fontSize=20,
        subtitleFontSize=16,
        offset=20,
    ), width=600, height=400
).configure_axis(
    grid=True,
    labelFontSize=12,
    titleFontSize=14
).configure_view(
    stroke="#ddd"
).configure_legend(
    labelFontSize=12
)

alt.Chart(...)

<span style='color:red'><b>Comments</b></span>: The average pricing for casual membership is around 3 times more expensive than the annual membership

- Berapa revenue dari setiap peminjaman pada tiap membership?

In [43]:
member_revenue = df.groupby(['member_casual', 'day_of_week'])['charge($)'].sum().reset_index()
member_revenue

member_casual  day_of_week     charge($)
0         casual            1  6.627460e+05
1         casual            2  6.481142e+05
2         casual            3  5.630418e+05
3         casual            4  6.399508e+05
4         casual            5  8.792184e+05
5         casual            6  1.351397e+06
6         casual            7  1.084441e+06
7     subscriber            1  3.778047e+05
8     subscriber            2  4.436720e+05
9     subscriber            3  4.368976e+05
10    subscriber            4  4.541126e+05
11    subscriber            5  4.558362e+05
12    subscriber            6  4.853786e+05
13    subscriber            7  3.959755e+05

In [44]:
bar = alt.Chart(member_revenue).mark_bar().encode(
    x=alt.X('member_casual', title='Membership Type', sort='-y', 
            axis=alt.Axis(labelAngle=0, labelFontSize=12)),
    y=alt.Y('charge($)', title='Revenue ($)', axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    tooltip=alt.Tooltip(['member_casual', 'charge($)']),
    color=alt.Color('member_casual', 
                    scale=alt.Scale(range=['orange', 'blue']), 
                    legend=None),
).properties(
    title=alt.Title(
        "Casual membership in total generates more revenue",
        subtitle='Total Revenue Based on Membership Types',
        anchor='start',
        fontSize=18,
        subtitleFontSize=14,
        offset=20,
    ), width=400, height=250
)

line = alt.Chart(member_revenue).mark_line().encode(
    alt.X('day_of_week:N', title='Day of Week', axis=alt.Axis(labelAngle=0, labelFontSize=12)),
    alt.Y('charge($)', title='Revenue ($)', axis=alt.Axis(labelFontSize=12)).axis(format='s'),
    color=alt.Color('member_casual')
).properties(
    title=alt.Title(
        "Revenue from casual riders peaked on weekends",
        subtitle='Total Revenue Based on Membership Types',
        anchor='start',
        fontSize=18,
        subtitleFontSize=14,
        offset=20,
    ), width=400, height=250
)

(bar | line).configure_axis(
    grid=True
).configure_view(
    stroke="#ddd"
)

alt.HConcatChart(...)

____

## `Question 3`: How digital media could affect their marketing tactics?
- Implement targeted ads based on top station names and time
    - In IG we can use geotagging-based ads
    - More efficient targetting the members and casual riders
    - Members for promotions whereas the casual riders for benefits of memberships

In [45]:
# Filter the DataFrame to include only casual riders
df_casual = df[df['member_casual'] == 'casual']

# Calculate the total number of casual people starting at each hour and start station
start_hour_station_counts_casual = df_casual.groupby(['start_hour', 'start_station_name']).size().reset_index(name='total_people')

# Filter stations with more than 1000 casual people per hour
start_hour_station_counts_casual_filtered = start_hour_station_counts_casual[start_hour_station_counts_casual['total_people'] > 1000]

# Sort the DataFrame by 'start_hour' for better visualization
start_hour_station_counts_casual_filtered = start_hour_station_counts_casual_filtered.sort_values(by=['start_hour', 'total_people'], ascending=[True, False])

# Create the Altair chart for casual riders
chart_casual = alt.Chart(start_hour_station_counts_casual_filtered).mark_bar().encode(
    x=alt.X('start_hour:O', axis=alt.Axis(labelAngle=0)),
    y='total_people:Q',
    color=alt.Color('start_station_name:N',legend=alt.Legend(title='Start Station Name')),
    tooltip=['start_hour:O', 'start_station_name:N', 'total_people:Q']
).properties(
    width=500,
    height=300,
    title=alt.Title('Casual riders',
                    anchor='start')
)

# Filter the DataFrame to include only member riders
df_member = df[df['member_casual'] == 'subscriber']

# Calculate the total number of member people starting at each hour and start station
start_hour_station_counts_member = df_member.groupby(['start_hour', 'start_station_name']).size().reset_index(name='total_people')

# Filter stations with more than 1000 member people per hour
start_hour_station_counts_member_filtered = start_hour_station_counts_member[start_hour_station_counts_member['total_people'] > 1000]

# Sort the DataFrame by 'start_hour' for better visualization
start_hour_station_counts_member_filtered = start_hour_station_counts_member_filtered.sort_values(by=['start_hour', 'total_people'], ascending=[True, False])

# Create the Altair chart for member riders
chart_member = alt.Chart(start_hour_station_counts_member_filtered).mark_bar().encode(
    x=alt.X('start_hour:O', axis=alt.Axis(labelAngle=0)),
    y='total_people:Q',
    color='start_station_name:N',
    tooltip=['start_hour:O', 'start_station_name:N', 'total_people:Q']
).properties(
    width=500,
    height=300,
    title=alt.Title('Subscribers',
                    anchor='start')
)

# Combine both charts side by side
combined_chart = alt.hconcat(chart_casual, chart_member)

combined_chart.properties(
    title=alt.Title('Where to put ads in Start Station',
                    fontSize=20,
                    offset=20)
)

alt.HConcatChart(...)

In [49]:
# Filter casual riders
df_casual = df[df['member_casual'] == 'casual'].copy()

# Extract the end hour
df_casual['end_hour'] = pd.to_datetime(df_casual['ended_at']).dt.hour

# Calculate the total number of casual riders
end_hour_station_counts = df_casual.groupby(['end_hour', 'end_station_name']).size().reset_index(name='total_people')

# Filter stations with more than 300 riders/h
end_hour_station_counts_filtered = end_hour_station_counts.loc[end_hour_station_counts['total_people'] > 1000]

end_hour_station_counts_filtered = end_hour_station_counts_filtered.sort_values(by=['end_hour', 'total_people'], ascending=[True, False])

# Filter the DataFrame to include only member riders
df_member = df[df['member_casual'] == 'subscriber'].copy()

# Extract the end hour from the 'ended_at' column
df_member['end_hour'] = pd.to_datetime(df_member['ended_at']).dt.hour

# Calculate the total number of member 
end_hour_station_counts_member = df_member.groupby(['end_hour', 'end_station_name']).size().reset_index(name='total_people')

# Filter stations with more than 1000 members per hour
end_hour_station_counts_member_filtered = end_hour_station_counts_member[end_hour_station_counts_member['total_people'] > 1000]

# Sort by 'end_hour' 
end_hour_station_counts_member_filtered = end_hour_station_counts_member_filtered.sort_values(by=['end_hour', 'total_people'], ascending=[True, False])

chart = alt.Chart(end_hour_station_counts_filtered).mark_bar().encode(
    x=alt.X('end_hour:O', axis=alt.Axis(labelAngle=0)),
    y='total_people:Q',
    color=alt.Color('end_station_name:N', legend=alt.Legend(title='Start Station Name')),
    tooltip=['end_hour:O', 'end_station_name:N', 'total_people:Q']
).properties(
    width=500,
    height=300,
    title=alt.Title('Casual members',
                    anchor='start')
)

chart_member = alt.Chart(end_hour_station_counts_member_filtered).mark_bar().encode(
    x=alt.X('end_hour:O', axis=alt.Axis(labelAngle=0)),
    y='total_people:Q',
    color='end_station_name:N',
    tooltip=['end_hour:O', 'end_station_name:N', 'total_people:Q']
).properties(
    width=500,
    height=300,
    title=alt.Title('Annual members',
                    anchor='start')
)

(chart | chart_member).properties(
    title=alt.Title('Where and when to show ads in end station',
                    anchor='start',
                    fontSize=20,
                    offset=20)
)


alt.HConcatChart(...)

- <span style='color:red'><b>Comment</b></span>: Annual members seem to be able to start using and return bikes in more flexible places as shown by the presence of `unknown` category compared to the casual riders.

In [47]:
# check total
customer = pd.crosstab(index=df['member_casual'], columns=df['customer_stage'], margins=True, margins_name='Total', normalize='index')
customer

customer_stage     aware  consider   convert
member_casual                               
casual          0.410903  0.589097  0.000000
subscriber      0.000000  0.000000  1.000000
Total           0.168143  0.241061  0.590796

In [48]:
customer = df.groupby('customer_stage').agg(func={'customer_stage':'count'}).reset_index(names='Category').rename(columns={'customer_stage':'Count'})
customer['Percent'] = customer['Count']/customer['Count'].sum()
customer

Category    Count   Percent
0     aware   370289  0.168143
1  consider   530870  0.241061
2   convert  1301066  0.590796

In [49]:
base = alt.Chart(customer).encode(
    alt.Theta("Percent").stack(True),
    alt.Color("Category").legend(),
).properties(
    title="Percentage of Customers Based on Stage",
    width=600, height=300
)


pie = base.mark_arc(outerRadius=120)
text = base.mark_text(radius=150, size=15).encode(
    text=alt.Text('Percent', format='.1%'),
)

pie + text

alt.LayerChart(...)

<span style='color:red'><b>Comments</b></span>:
- The majority of customers have been converted as annual members
- As a result, marketing team can focus more on those who are still considering or am aware.

<h1 style='text-align:center'><i><b>---END---</b></i></h1>